### PR TITLE
Groups added to the 'Other groups' field are lost on submit

### PR DESCRIPTION
**Pull request by pfrenssen**
I noticed that the 'Other Groups' field is not working at all. There are two bugs there:
- If the 'Group Audience' field is left empty, a target reference with ID `NULL` is added to the list.
- The wrong element in the `$form` array is inspected when looking for the field.

A third bug is preventing the tests from completing:
- The config schema for the complex widget is missing.

This has been split off from the larger work of converting the OG Complex widget tests in #140, so that this important fix can go through without waiting for that mammoth task to complete. I have also cherry-picked the two commits from #147 that allow us to run functional tests on Travis.

### CODE REVIEW

**Comment by pfrenssen** [comment]
Above test fails correctly because it detects that it is impossible to add groups through the 'Other groups' field.

**Comment by pfrenssen** [comment]
This fixes it and should turn the test green. Note that there is still a bug left: it now hardcodes the field name, so it only works for one group audience field. This will be tackled when I get to `OgComplexWidgetTest::testMultipleFields()`.

File: src/Plugin/Field/FieldWidget/OgComplex.php [hunk 1]
```diff
@@ -306,10 +306,14 @@ public function otherGroupsSingle($delta, EntityInterface $entity = NULL, $weigh
...
-    $parent_values = $values;
+    // Remove empty values. The form fields may be empty.
+    $values = array_filter($values, function ($item) {
+      return !empty($item['target_id']);
+    });
 
     // Get the groups from the other groups widget.
-    foreach ($form['other_groups'] as $key => $value) {
+    // @todo Don't hardcode 'og_group_ref' - there can be multiple fields.
+    foreach ($form['og_group_ref']['other_groups'] as $key => $value) {
```
Comment (damiankloip) [inline src/Plugin/Field/FieldWidget/OgComplex.php]: `$this->fieldDefinition->getName()` :+1:

**Comment by pfrenssen** [comment]
Thanks @damiankloip! Much obliged :)

File: tests/src/Functional/OgComplexWidgetTest.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    // group content type.
+    $this->createContentType(['type' => 'post']);
+    Og::createField(OgGroupAudienceHelper::DEFAULT_FIELD, 'node', 'post');
+
+    // Make the group audience field visible in the default form display.
+    // @todo Remove this once issue #144 is in.
+    // @see https://github.com/amitaibu/og/issues/144
+    /** @var EntityFormDisplayInterface $form_display */
+    $form_display = \Drupal::entityTypeManager()->getStorage('entity_form_display')->load('node.post.default');
+    $widget = $form_display->getComponent('og_group_ref');
```
Comment (amitaibu) [inline tests/src/Functional/OgComplexWidgetTest.php]: as indicated in the issue, this one troubles me a bit (even if it's an intermediate step). I'd be happy to see the next step to make this work for all audience fields

Comment (pfrenssen) [inline tests/src/Functional/OgComplexWidgetTest.php]: Ok let's do #144 first, it should be pretty trivial.

**Comment by pfrenssen** [comment]
The tests started failing after merging in the master branch. It was due to the OgStandardReferenceItem that was added in #128 - it was using the wrong validator, causing the following error when saving a node that references a group:

```
Error: Call to undefined method Drupal\Core\TypedData\Plugin\DataType\IntegerData::isEmpty() in Drupal\Core\Entity\Plugin\Validation\Constraint\ValidReferenceConstraintValidator->validate() (line 78 of core/lib/Drupal/Core/Entity/Plugin/Validation/Constraint/ValidReferenceConstraintValidator.php).
```

This test was actually detecting that already :)

@amitaibu, I have fixed the visibility of the group audience field in #153, this is merged into this issue. The fix to make it work with all audience fields has also been applied. Is there anything else that needs to be done here?

**Comment by pfrenssen** [comment]
I created a separate issue for the fatal error when saving group content, so that it is more easily discoverable if someone runs into this problem: https://github.com/amitaibu/og/issues/154.

**Comment by RoySegall** [comment]
@amitaibu Checked it and the other groups widget works very good.

**Comment by amitaibu** [comment]
Merged, thanks!

**Pull request merged**